File: lerobot-tagged/python/src/lerobot_tagged/recorder.py
"""recorder.py — Record tag detections and replay in MuJoCo.

Example
-------
>>> import cv2
>>> from lerobot_tagged import Tag, Detector, Recorder
>>>
>>> detector = Detector(
...     tags=[Tag(101, offset=(-5.0, 1.0), width_mm=25, height_mm=75, depth_mm=25)],
...     H=H, scale=scale, inset_mm=16,
... )
>>>
>>> cap = cv2.VideoCapture(0)
>>> recorder = Recorder(cap, detector)
>>> recorder.record()
>>> # ... do other things while recording ...
>>> recorder.stop_recording()
>>>
>>> for frame in recorder.frames:
...     print(frame.timestamp, frame.tag_locations)
>>>
>>> recorder.output_mujoco_recording("scene.mp4")
"""

from __future__ import annotations

import importlib.resources
import threading
import time
from dataclasses import dataclass, field
from pathlib import Path
from typing import Dict, List, Optional, Tuple

import cv2
import numpy as np

from .detection import Detector, Tag, TagLocation


# Dummy joint names for the SO-101 robot arm.
# These will be populated with real values in a future update.
_DUMMY_JOINTS: Dict[str, float] = {
    "shoulder_pan":   0.0,
    "shoulder_lift":  0.0,
    "elbow_flex":     0.0,
    "wrist_flex":     0.0,
    "wrist_roll":     0.0,
    "gripper":        0.0,
}


@dataclass
class Frame:
    """A single recorded frame.

    Attributes
    ----------
    timestamp:
        Seconds elapsed since ``record()`` was called.
    tag_locations:
        All detected tag locations in this frame.
    joint_positions:
        Robot arm joint angles in radians, keyed by joint name.
        Currently returns dummy zero values; will be populated from
        real hardware in a future update.
    end_effector:
        (x, y, z) end-effector position in world metres.
        Currently a dummy value; will be computed from joint positions.
    """
    timestamp: float
    tag_locations: List[TagLocation]
    joint_positions: Dict[str, float]
    end_effector: Tuple[float, float, float]


class Recorder:
    """Record tag detections from a camera and replay in MuJoCo.

    Parameters
    ----------
    cap:
        An open ``cv2.VideoCapture`` instance.
    detector:
        A configured :class:`~lerobot_tagged.Detector` instance.

    Example
    -------
    >>> recorder = Recorder(cap, detector)
    >>> recorder.record()        # starts background thread, non-blocking
    >>> time.sleep(5)
    >>> recorder.stop_recording()
    >>> print(len(recorder.frames), "frames recorded")
    >>> recorder.output_mujoco_recording("output.mp4")
    """

    def __init__(self, cap: cv2.VideoCapture, detector: Detector) -> None:
        self._cap = cap
        self._detector = detector
        self._frames: List[Frame] = []
        self._thread: Optional[threading.Thread] = None
        self._stop_event = threading.Event()

    def record(self) -> None:
        """Start recording in a background thread (non-blocking).

        Raises
        ------
        RuntimeError
            If recording is already in progress.
        """
        if self._thread is not None and self._thread.is_alive():
            raise RuntimeError("Recording is already in progress. Call stop_recording() first.")

        self._frames = []
        self._stop_event.clear()
        self._thread = threading.Thread(target=self._capture_loop, daemon=True)
        self._thread.start()

    def stop_recording(self) -> None:
        """Signal the recording thread to stop and wait for it to finish."""
        self._stop_event.set()
        if self._thread is not None:
            self._thread.join()
            self._thread = None

    @property
    def frames(self) -> List[Frame]:
        """All recorded frames, in chronological order.

        Safe to read after :meth:`stop_recording` is called.
        """
        return list(self._frames)

    def output_mujoco_recording(
        self,
        output_path: str = "mujoco_recording.mp4",
        xml_path: Optional[str] = None,
        camera: str = "topdown",
        width: int = 600,
        height: int = 600,
        fps: float = 30.0,
    ) -> None:
        """Render the recorded frames in MuJoCo and write to a video file.

        Each frame positions the tracked objects in the scene according to
        their detected ``pos`` coordinates, then renders
        from the specified camera.

        Board interior mm coordinates are mapped to MuJoCo world metres with
        X → X, Y → Y (both divided by 1000). Z is set to half the object's
        ``height_mm`` so it rests on the floor. Override with a custom
        ``xml_path`` to use your own coordinate system.

        Parameters
        ----------
        output_path:
            Path for the output video file (mp4).
        xml_path:
            Path to a MuJoCo XML scene. When omitted, an auto-generated
            scene is used that includes a box body for each tracked tag
            (sized from ``width_mm``, ``height_mm``, ``depth_mm``).
        camera:
            Name of the MuJoCo camera to render from.
        width, height:
            Output video dimensions in pixels.
        fps:
            Frames per second for the output video.
        """
        try:
            import mujoco
        except ImportError as e:
            raise ImportError(
                "mujoco is required for output_mujoco_recording. "
                'Install it with: pip install "lerobot-tagged[mujoco]"'
            ) from e

        if not self._frames:
            raise RuntimeError("No frames recorded. Call record() and stop_recording() first.")

        # Build or load the MuJoCo model
        if xml_path is not None:
            model = mujoco.MjModel.from_xml_path(xml_path)
        else:
            xml = self._generate_scene_xml()
            model = mujoco.MjModel.from_xml_string(xml)

        data = mujoco.MjData(model)
        mujoco.mj_forward(model, data)

        renderer = mujoco.Renderer(model, height=height, width=width)

        fourcc = cv2.VideoWriter_fourcc(*"mp4v")
        writer = cv2.VideoWriter(output_path, fourcc, fps, (width, height))

        tags_by_id = self._detector.tags

        for frame in self._frames:
            # Position each detected object
            for loc in frame.tag_locations:
                tag = tags_by_id.get(loc.tag_id)
                body_name = f"object_{loc.tag_id}"

                body_id = mujoco.mj_name2id(model, mujoco.mjtObj.mjOBJ_BODY, body_name)
                if body_id < 0:
                    continue

                jnt_adr = model.body_jntadr[body_id]
                if jnt_adr < 0:
                    continue

                x_mm, y_mm = loc.pos
                x = x_mm / 1000.0
                y = y_mm / 1000.0
                z = (tag.height_mm or 25) / 2000.0  # rest on floor

                qpos_adr = model.jnt_qposadr[jnt_adr]
                data.qpos[qpos_adr:qpos_adr + 3] = [x, y, z]
                data.qpos[qpos_adr + 3:qpos_adr + 7] = [1, 0, 0, 0]

            mujoco.mj_forward(model, data)
            renderer.update_scene(data, camera=camera)
            rgb = renderer.render().copy()
            bgr = cv2.cvtColor(rgb, cv2.COLOR_RGB2BGR)
            writer.write(bgr)

        renderer.close()
        writer.release()

    # ── internals ────────────────────────────────────────────────────────────

    def _capture_loop(self) -> None:
        t0 = time.monotonic()
        while not self._stop_event.is_set():
            ret, frame_img = self._cap.read()
            if not ret:
                break

            timestamp = time.monotonic() - t0
            tag_locations = self._detector.find(frame_img)

            self._frames.append(Frame(
                timestamp=timestamp,
                tag_locations=tag_locations,
                joint_positions=dict(_DUMMY_JOINTS),
                end_effector=(0.0, 0.0, 0.0),
            ))

    def _generate_scene_xml(self) -> str:
        """Auto-generate a MuJoCo XML scene with one box body per tracked tag."""
        bodies = []
        for tag in self._detector.tags.values():
            w = (tag.width_mm  or 25) / 2000.0  # half-size in metres
            h = (tag.height_mm or 25) / 2000.0
            d = (tag.depth_mm  or 25) / 2000.0
            bodies.append(
                f'    <body name="object_{tag.id}" pos="0 0 {h}">\n'
                f'      <freejoint/>\n'
                f'      <geom type="box" size="{w} {d} {h}" rgba="0.8 0.2 0.2 1"/>\n'
                f'    </body>'
            )

        bodies_xml = "\n".join(bodies)
        return f"""<mujoco model="lerobot-tagged-scene">
  <option timestep="0.002"/>
  <visual>
    <headlight ambient="0.4 0.4 0.4" diffuse="0.6 0.6 0.6"/>
  </visual>
  <worldbody>
    <light pos="0 0 1.5" dir="0 0 -1" diffuse="0.7 0.7 0.7" specular="0.1 0.1 0.1"/>
    <geom name="floor" type="plane" size="1 1 0.05" rgba="0.85 0.85 0.85 1"/>
    <camera name="topdown" pos="0 0 0.8" euler="0 0 0"/>
{bodies_xml}
  </worldbody>
</mujoco>"""

File: lerobot-tagged/js/src/recorder.ts
/**
 * recorder.ts — Record tag detections from a video source and replay in MuJoCo.
 *
 * @example
 * ```ts
 * import { Tag, Detector, Recorder } from "lerobot-tagged";
 *
 * const detector = new Detector(
 *   [new Tag(101, { offset: [-5, 1], widthMm: 25, heightMm: 75, depthMm: 25 })],
 *   { H, scale, insetMm: 16 }
 * );
 *
 * const video = document.querySelector("video")!;
 * const recorder = new Recorder(video, detector);
 *
 * recorder.record();
 * // ... after some time ...
 * recorder.stopRecording();
 *
 * console.log(recorder.frames);
 * const renderedFrames = await recorder.outputMujocoRecording(mj);
 * ```
 */

import { Detector, TagLocation } from "./detection.js";

/** Dummy SO-101 joint names. Will be populated from hardware in a future update. */
const DUMMY_JOINTS: Record<string, number> = {
  shoulder_pan:  0,
  shoulder_lift: 0,
  elbow_flex:    0,
  wrist_flex:    0,
  wrist_roll:    0,
  gripper:       0,
};

/** A single recorded frame. */
export interface Frame {
  /** Seconds elapsed since {@link Recorder.record} was called. */
  timestamp: number;
  /** All detected tag locations in this frame. */
  tagLocations: TagLocation[];
  /**
   * Robot arm joint angles in radians, keyed by joint name.
   * Currently dummy zero values; will be populated from hardware in a future update.
   */
  jointPositions: Record<string, number>;
  /**
   * End-effector position in world metres [x, y, z].
   * Currently a dummy value; will be computed from joint positions in a future update.
   */
  endEffector: [number, number, number];
}

/**
 * Records tag detections from an HTMLVideoElement and replays them in MuJoCo.
 *
 * @example
 * ```ts
 * const recorder = new Recorder(videoElement, detector);
 * recorder.record();
 * await new Promise(r => setTimeout(r, 5000));
 * recorder.stopRecording();
 * const rgbFrames = await recorder.outputMujocoRecording(mj);
 * ```
 */
export class Recorder {
  private readonly video: HTMLVideoElement;
  private readonly detector: Detector;
  private readonly canvas: HTMLCanvasElement;
  private readonly ctx: CanvasRenderingContext2D;

  private _frames: Frame[] = [];
  private _rafHandle: number | null = null;
  private _startTime: number = 0;

  constructor(video: HTMLVideoElement, detector: Detector) {
    this.video = video;
    this.detector = detector;
    this.canvas = document.createElement("canvas");
    this.ctx = this.canvas.getContext("2d")!;
  }

  /** Start recording (non-blocking — runs on requestAnimationFrame). */
  record(): void {
    if (this._rafHandle !== null) {
      throw new Error("Recording is already in progress. Call stopRecording() first.");
    }
    this._frames = [];
    this._startTime = performance.now();
    this._loop();
  }

  /** Stop recording. Safe to call even if not currently recording. */
  stopRecording(): void {
    if (this._rafHandle !== null) {
      cancelAnimationFrame(this._rafHandle);
      this._rafHandle = null;
    }
  }

  /** All recorded frames in chronological order. */
  get frames(): Frame[] {
    return [...this._frames];
  }

  /**
   * Render all recorded frames in a MuJoCo scene and return the RGB frames.
   *
   * Each frame positions tracked objects according to their detected
   * ``pos`` coordinates (board interior mm → world metres:
   * X→X, Y→Y, both ÷ 1000; Z = half object height).
   *
   * In Node with the `fs` module available, frames are written as individual
   * PNGs to ``outputDir``. In both environments the rendered RGB Uint8Arrays
   * are returned for further processing.
   *
   * @param mj - Initialised MuJoCo WASM module.
   * @param xml - MuJoCo XML string. Defaults to an auto-generated scene with
   *   one box body per tracked tag.
   * @param outputDir - Directory to write PNG frames (Node only). Defaults to
   *   ``"mujoco_recording/"``
   * @param camera - MuJoCo camera name to render from.
   * @param width - Render width in pixels.
   * @param height - Render height in pixels.
   * @returns Array of rendered RGB frames (Uint8Array, width × height × 3).
   */
  async outputMujocoRecording(
    mj: any,
    xml?: string,
    outputDir: string = "mujoco_recording",
    camera: string = "topdown",
    width: number = 600,
    height: number = 600,
  ): Promise<Uint8Array[]> {
    if (this._frames.length === 0) {
      throw new Error("No frames recorded. Call record() and stopRecording() first.");
    }

    const sceneXml = xml ?? this._generateSceneXml();
    const model = mj.MjModel.from_xml_string(sceneXml);
    const data = new mj.MjData(model);
    mj.mj_forward(model, data);

    const renderer = new mj.Renderer(model, height, width);
    const rendered: Uint8Array[] = [];

    for (const frame of this._frames) {
      for (const loc of frame.tagLocations) {
        const tag = this.detector.tags.get(loc.tagId);
        const bodyName = `object_${loc.tagId}`;
        const bodyId = mj.mj_name2id(model, mj.mjtObj.mjOBJ_BODY.value, bodyName);
        if (bodyId < 0) continue;

        const jntAdr = model.body_jntadr[bodyId];
        if (jntAdr < 0) continue;

        const [xMm, yMm] = loc.pos;
        const heightMm = tag?.heightMm ?? 25;
        const qposAdr = model.jnt_qposadr[jntAdr];
        data.qpos[qposAdr]     = xMm / 1000;
        data.qpos[qposAdr + 1] = yMm / 1000;
        data.qpos[qposAdr + 2] = heightMm / 2000;
        data.qpos[qposAdr + 3] = 1; // quaternion w
        data.qpos[qposAdr + 4] = 0;
        data.qpos[qposAdr + 5] = 0;
        data.qpos[qposAdr + 6] = 0;
      }

      mj.mj_forward(model, data);
      renderer.update_scene(data, camera);
      rendered.push(new Uint8Array(renderer.render()));
    }

    renderer.free();

    // Write to disk in Node
    try {
      const { writeFileSync, mkdirSync } = await import("fs");
      mkdirSync(outputDir, { recursive: true });
      rendered.forEach((frame, i) => {
        const padded = String(i).padStart(6, "0");
        writeFileSync(`${outputDir}/frame_${padded}.png`, frame);
      });
    } catch {
      // Not in Node — skip file writing
    }

    return rendered;
  }

  // ── internals ──────────────────────────────────────────────────────────────

  private _loop(): void {
    this._rafHandle = requestAnimationFrame(async () => {
      if (this._rafHandle === null) return; // stopped

      this.canvas.width = this.video.videoWidth;
      this.canvas.height = this.video.videoHeight;
      this.ctx.drawImage(this.video, 0, 0);
      const imageData = this.ctx.getImageData(0, 0, this.canvas.width, this.canvas.height);

      const timestamp = (performance.now() - this._startTime) / 1000;
      const tagLocations = await this.detector.find(imageData);

      this._frames.push({
        timestamp,
        tagLocations,
        jointPositions: { ...DUMMY_JOINTS },
        endEffector: [0, 0, 0],
      });

      this._loop();
    });
  }

  private _generateSceneXml(): string {
    const bodies = [...this.detector.tags.values()].map(tag => {
      const w = (tag.widthMm  ?? 25) / 2000;
      const h = (tag.heightMm ?? 25) / 2000;
      const d = (tag.depthMm  ?? 25) / 2000;
      return `    <body name="object_${tag.id}" pos="0 0 ${h}">
      <freejoint/>
      <geom type="box" size="${w} ${d} ${h}" rgba="0.8 0.2 0.2 1"/>
    </body>`;
    }).join("\n");

    return `<mujoco model="lerobot-tagged-scene">
  <option timestep="0.002"/>
  <visual>
    <headlight ambient="0.4 0.4 0.4" diffuse="0.6 0.6 0.6"/>
  </visual>
  <worldbody>
    <light pos="0 0 1.5" dir="0 0 -1" diffuse="0.7 0.7 0.7" specular="0.1 0.1 0.1"/>
    <geom name="floor" type="plane" size="1 1 0.05" rgba="0.85 0.85 0.85 1"/>
    <camera name="topdown" pos="0 0 0.8" euler="0 0 0"/>
${bodies}
  </worldbody>
</mujoco>`;
  }
}
